### DISCUSSION TITLE
RabbitMQ upgrade from 3.9.29 to 3.10.24 failed and got queue index directory not found error

### DISCUSSION THREAD

**Discussion by CassieZAHNG** [rabbitmq/rabbitmq-server#8663] [Questions] [answered]
Hey,

I am trying to upgrade RabbitMQ in AKS clusters (three pods per cluster) with rolling upgrade strategy. The old container image is `rabbitmq:3.9-management`, now try to upgrade to `rabbitmq:3.10-management`. 

Have checked all required feature flags are enabled as required from [RabbitMQ 3.10.0](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.10.0). All clusters using the same rabbit configurations, but 2 clusters successfully upgraded while 2 got failed, not sure what could make it different.

Attached logs for failed cluster, it failed at step of moving messages to per-vhost message store:
![image](https://github.com/rabbitmq/rabbitmq-server/assets/34259576/277a254e-dfd8-4c9f-bcd5-e3a21fff047e)

For the successfully upgrade one, it says no durable queue is found so it skips message store migration:
![image](https://github.com/rabbitmq/rabbitmq-server/assets/34259576/f2f6899f-885b-42dc-b451-a8ab1e0d877d)

Have done additional testing, when I tried to downgrade back to 3.9.29, all pods running again, but if re-upgrade pods then it will fail with the same error.

From my understanding, RabbitMQ will take care of migrating durable queue when upgrading, does not require additional steps from users? 

Please let me know if require further information, super appreciate your help!

**Comment by michaelklishin** [comment]
Somehow this node ended up without an index directory for one of its queues and a (3.7 era) migration function assumes that cannot be the case. We cannot tell why that happened.

You can delete the directory in question to let the node proceed.

RabbitMQ does not support downgrades.

**Reply by CassieZAHNG** [reply]
Thanks for replying  @michaelklishin ! Got some follow up questions regarding delete the queue index directory. Where is RabbitMQ queue index directory located? I have shelled into the pod with the path indicated in the error msg but have not found queue subdirectory. Have checked successfully upgrade clusters, they also do not have queue directory, does that something should exist?
![image](https://github.com/rabbitmq/rabbitmq-server/assets/34259576/85e7f82e-cac2-4fe0-94ae-42e4f2b32f76)

How queue and queue index directory work? Does delete queue will not trigger delete index directory and associated cache automatically? I have tried to delete a queue from Rabbit UI management, but the queue index directory error still there when upgrade, seems we need to perform additional step to clean up all cache and directory.

Also, would like to know if there is a recommended way to delete it (e.g., some sort of RabbitMQ maintenance strategy to put the queue/ queue metadata cache/ queue index directory in a clean state)? Or do we need to delete it manually? If we have to go with manual way, how do we make sure we have cleaned up all queue cache completely?

I googled a bit, someone mentioned can perform following cmd to clean up all queuing system, but sounds too risky to me as it returns RabbitMQ to default statue and removing all settings like VHost and users, etc. So would like to hear your thoughts first:

```
rabbitmqctl stop_app
rabbitmqctl reset
rabbitmqctl start_app
```

**Reply by mkuratczyk** [reply]
`rabbitmqctl reset` is not the answer indeed - you'd lose all node's state. 
You have the path to the queue folder in the error that was logged.

**Reply by CassieZAHNG** [reply]
I have tried to access to the logged path but cannot see subdirectory named "queue" under mnesia/rabbit@node-host-name. So instead of deleting queue folder, I have tried to delete its parent folder: "/var/lib/rabbitmq/mnesia/rabbit@rabbitmq-host-name/'

Then I performed an upgrade but still got the same error msg. Is that because stateful set will automatically create a new queue with pod restart when perform an upgrade? If that the case, what could be the solution?

Thanks for the reply!

**Reply by mkuratczyk** [reply]
if you deleted `/var/lib/rabbitmq/mnesia/rabbit@rabbitmq-host-name/` then you already deleted most this node's state. I'm not sure what to suggest here, at this point your node is probably completely broken. If think you can try `rabbitmqctl reset`

**Reply by CassieZAHNG** [reply]
Sure. Tried rabbitmqctl reset still no luck got the same error. But one thing I've noticed is, if the version is 3.9.29, even if got the same error but rabbitmq will create index from scratch as following logs shows:

```
Message store "rabbit@rabbitmq-0.rabbitmq-headless.rabbit.svc.cluster.local/msg_store_persistent": using rabbit_msg_store_ets_index to provide index
Message store "rabbit@rabbitmq-0.rabbitmq-headless.rabbit.svc.cluster.local/msg_store_persistent": rebuilding indices from scratch
```

While if the version is 3.10.24, once the "Queue index directory cannot be found", it will immediately crash instead of recreating. Does 3.9.x to 3.10.x has some breaking changes associated with classic durable queue index?

**Reply by mkuratczyk** [reply]
I don't recall any massive changes, but you'd need to go through the changelog/commits yourself. It's been more than a year since 3.10 was released.

**Reply by CassieZAHNG** [reply]
yep, I have checked the release notes, but have not found index issue same as the one we met. Will continue doing some trials and errors, thanks for the help!

**Reply by CassieZAHNG** [reply]
Hey @mkuratczyk, got a follow up question regarding previous issue:

We are trying to upgrade from 3.9 to 3.10. And our msg stores are already in "/var/lib/rabbitmq/mnesia/rabbit@rabbit-hostname/msg_stores/vhosts" format. So ideally shouldn't the "move_messages_to_vhost_store" migration function returned nothing to migrate and skipped migration? Is that understanding correct?

**Reply by mkuratczyk** [reply, marked as answer]
Yes, I believe that's correct. You can follow the code here
https://github.com/rabbitmq/rabbitmq-server/blob/v3.10.x/deps/rabbit/src/rabbit_variable_queue.erl#L3398-L3444

**Reply by CassieZAHNG** [reply]
Checked the above link only pre-3.7.0 version required to msg_store migrate, not sure why 3.10 still got the issue. Now upgrade to 3.12, the issues are gone. Thanks for the prompt help!
